near-plugins-derive/src/upgradable.rs

```diff
@@ -50,9 +121,64 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
 
             #[#cratename::only(owner)]
             fn up_deploy_code(&mut self) -> near_sdk::Promise {
+                let staging_timestamp = self.up_get_timestamp(__UpgradableStorageKey::StagingTimestamp)
+                    .unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: staging timestamp isn't set"));
+
+                if near_sdk::env::block_timestamp() < staging_timestamp {
+                    near_sdk::env::panic_str(
+                        format!(
+                            "Upgradable: Deploy code too early: staging ends on {}",
+                            staging_timestamp
+                        )
+                        .as_str(),
+                    );
+                }
+
                 near_sdk::Promise::new(near_sdk::env::current_account_id())
                     .deploy_contract(self.up_staged_code().unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: No staged code")))
             }
+
+            #[#cratename::only(owner)]
+            fn up_init_staging_duration(&self, staging_duration: near_sdk::Duration) {
+                near_sdk::require!(self.up_get_duration(__UpgradableStorageKey::StagingDuration).is_none(), "Upgradable: staging duration was already initialized");
+                self.up_set_staging_duration_unchecked(staging_duration);
+            }
+
+            #[#cratename::only(owner)]
+            fn up_stage_update_staging_duration(&self, staging_duration: Option<near_sdk::Duration>) {
+                if let Some(staging_duration) = staging_duration {
+                    let current_staging_duration = self.up_get_duration(__UpgradableStorageKey::StagingDuration)
+                        .unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: staging duration isn't initialized"));
+
+                    let staging_duration_timestamp = near_sdk::env::block_timestamp() + current_staging_duration;
+                    self.up_set_timestamp(__UpgradableStorageKey::StagingTimestamp, staging_duration_timestamp);
+                    self.up_set_duration(__UpgradableStorageKey::UpdateStagingDuration, staging_duration);
+                } else {
+                    near_sdk::env::storage_remove(&self.up_storage_key(__UpgradableStorageKey::UpdateStagingDuration));
```

Comment:
If the deletion of the delay key is immediate, couldn't the owner work around the delay period in updating the delay key by calling `up_stage_update_staging_duration(None)` and then `up_init_staging_duration(new_duration)`? This seems like a loop hole we do not want. I think either (a) deleting the key should be subject to the same delay as setting a new value or (b) it should not be allowed to delete the delay key at all (setting a delay of 0 would be sort of like removing the delay functionality).

Reply:
Initially I was thinking the same, though seems like `UpdateStagingDuration` holds the value that shall become the new delay. Then removing it would have the effect of aborting a staged update of the delay. If that's the case, I think the name `UpdateStagingDuration` is misleading.

Reply:
Ok, maybe we could rename it to  `StagedStagingDuration` or `NewStagingDuration`?

Reply:
`NewStagingDuration` sounds good to me. Though before renaming it, maybe let’s make sure we’re all on the same page regarding the delay update mechanism (see comments below).

Reply:
Renamed in c1654aea3df463524579d02fa8a091b8c59fb017